File: python/Losses.py
from abc import abstractmethod
import numpy as np
from abc import ABC

class Loss(ABC):

    @abstractmethod
    def loss():
        pass

    def calculateLoss(self, y_true, y_pred):
        return np.mean(self.loss(y_true, y_pred))


class CategoricalCrossentropy(Loss):

    def loss(self, y_true, y_pred):
        """Calculates the loss. Expects y_true to be one-hot encoded."""
        return -np.log(np.sum(np.multiply(y_true, y_pred), axis=1))

class SparseCategoricalCrossentropy(Loss):

    def loss(self, y_true, y_pred):
        """Calculates the loss. Expects y_true to be a 1D array of integers """
        return -np.log(y_pred[range(len(y_true)), y_true])

File: c++/losses.cpp
#include "matrix.h"
#include "losses.h"

double Loss::calculateLoss(matrix::Matrix<double> y_true, 
	    matrix::Matrix<double> y_pred) {
    return matrix::sum(loss(y_true, y_pred), 1, false)[0][0] / y_true.rows; 
}

matrix::Matrix<double> CategoricalCrossentropy::loss(matrix::Matrix<double> y_true, 
	    matrix::Matrix<double> y_pred) {
    // Expects y_true to be one hot encoded
    return matrix::mul_const(matrix::log(matrix::sum(matrix::mul(y_true, 
			y_pred))), (double) -1.0);
}

matrix::Matrix<double> SparseCategoricalCrossentropy::loss(matrix::Matrix<double> y_true, 
	    matrix::Matrix<double> y_pred) {
    // Expects y_true to be 1D array of integers. 
    // NOT IMPLEMENTED CORRECTLY YET.
    // NEEDS TO BE -np.log(y_pred[range(len(y_true)), y_true]):wq 
    return matrix::mul_const(matrix::log(y_true), (double) -1.0);
}
